java/flight/flight-sql/src/test/java/org/apache/arrow/flight/sql/example/FlightSqlExample.java

```diff
@@ -0,0 +1,1733 @@
+      throws SQLException {
+    saveToVectors(vectorToColumnName, data, false);
+  }
+
+  private static VectorSchemaRoot getTableTypesRoot(final ResultSet data, final BufferAllocator allocator)
+      throws SQLException {
+    return getRoot(data, allocator, "table_type", "TABLE_TYPE");
+  }
+
+  private static VectorSchemaRoot getCatalogsRoot(final ResultSet data, final BufferAllocator allocator)
+      throws SQLException {
+    return getRoot(data, allocator, "catalog_name", "TABLE_CATALOG");
+  }
+
+  private static VectorSchemaRoot getRoot(final ResultSet data, final BufferAllocator allocator,
+                                          final String fieldVectorName, final String columnName)
+      throws SQLException {
+    final VarCharVector dataVector = new VarCharVector(fieldVectorName, allocator);
+    saveToVectors(ImmutableMap.of(dataVector, columnName), data);
+    final int rows = dataVector.getValueCount();
+    dataVector.setValueCount(rows);
+    return new VectorSchemaRoot(singletonList(dataVector));
+  }
+
+  private static VectorSchemaRoot getTablesRoot(final DatabaseMetaData databaseMetaData,
+                                                final BufferAllocator allocator,
+                                                final boolean includeSchema,
+                                                final String catalog,
+                                                final String schemaFilterPattern,
+                                                final String tableFilterPattern,
+                                                final String... tableTypes)
+      throws SQLException, IOException {
+    /*
+     * TODO Fix DerbyDB inconsistency if possible.
+     * During the early development of this prototype, an inconsistency has been found in the database
+     * used for this demonstration; as DerbyDB does not operate with the concept of catalogs, fetching
+     * the catalog name for a given table from `DatabaseMetadata#getColumns` and `DatabaseMetadata#getSchemas`
+     * returns null, as expected. However, the inconsistency lies in the fact that accessing the same
+     * information -- that is, the catalog name for a given table -- from `DatabaseMetadata#getSchemas`
+     * returns an empty String.The temporary workaround for this was making sure we convert the empty Strings
+     * to null using `com.google.common.base.Strings#emptyToNull`.
+     */
+    Objects.requireNonNull(allocator, "BufferAllocator cannot be null.");
+    final VarCharVector catalogNameVector = new VarCharVector("catalog_name", allocator);
+    final VarCharVector schemaNameVector = new VarCharVector("schema_name", allocator);
+    final VarCharVector tableNameVector = new VarCharVector("table_name", allocator);
+    final VarCharVector tableTypeVector = new VarCharVector("table_type", allocator);
+
+    final List<FieldVector> vectors =
+        new ArrayList<>(
```

Comment:
why wrapping the immutable list into an array list? Is vectors supposed to be modifiable?